File: pascal/2015/day08b/day08b.pas
PROGRAM Day08b(Input, Output);

VAR
    FileName : PChar;
    Res      : Integer;

PROCEDURE Usage();
VAR
    ProgName : PChar;
BEGIN
    ProgName := argv[0];
    WriteLn('usage: ', ProgName, ' <input file>');
    Halt(1)
END;

FUNCTION Process(FileName : PChar) : Integer;
VAR
    InFile          : Text;
    Res             : Integer;
    LineStr         : String;
    CodeLen, EncLen : Integer;
    I               : Integer;

BEGIN
    Res := 0;
    {$I-}
    Assign(InFile, FileName);
    Reset(InFile);
    IF IOResult <> 0 THEN
    BEGIN
        WriteLn(StdErr, 'error when opening file: ', FileName);
        Halt(1)
    END;
    {$I+}
    WHILE NOT EOF(InFile) DO
    BEGIN
        ReadLn(InFile, LineStr);
        CodeLen := Length(LineStr);
        EncLen := 0;
        FOR I := 1 TO Length(LineStr) DO
            CASE LineStr[I] OF
                '\', '"' : EncLen := EncLen + 2;
            ELSE
                EncLen := EncLen + 1;
            END;
        Res := Res + 2 + (EncLen - CodeLen)
    END;
    Close(InFile);
    Process := Res
END;

BEGIN
    IF argc < 2 THEN
        Usage();
    FileName := argv[1];
    Res := Process(FileName);
    WriteLn('result = ', Res)
END.

File: c++/2015/day08b/day08b.cpp
#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>

void usage(std::string progname) {
  std::cerr << "usage: " << progname << " <input file>" << std::endl;
  std::exit(1);
}

uint32_t process(std::string filename) {
  uint32_t result = 0;
  std::ifstream infile(filename);
  std::string line;
  while (std::getline(infile, line)) {
    uint32_t code_len = line.length();
    uint32_t enc_len = 0;
    for (auto i = 0; i < line.length(); i++) {
      switch (line[i]) {
        case '\\':
        case '\"':
          enc_len += 2;
          break;
        default:
          enc_len++;
      }
    }
    result += 2 + (enc_len - code_len);
  }
  return result;
}

int main(int argc, char *argv[]) {
  std::string progname(argv[0]);
  if (argc < 2) {
    usage(progname);
  }
  std::string filename(argv[1]);
  uint32_t result = process(filename);
  std::cout << "result = " << result << std::endl;
  return 0;
}
